### DISCUSSION TITLE
Sunshine duration computation

### DISCUSSION THREAD

**Discussion by guidocioni** [open-meteo/open-meteo#1024] [Q&A]
So on my meteogram page I'm trying to put an estimation of daily sunshine duration. My idea was to reproduce something similar to what meteoblue has
![Screenshot_20240916-124927](https://github.com/user-attachments/assets/9f021b11-476b-40ee-a5b0-2e8a229a7f89)

I'm using the daily values from the normal forecast endpoint. If i understood well this is just computed based on surface irradiance, so any sunshine duration already present in the model output (some have that variable available) is ignored.

When trying to make the same picture on my app I get values that are much higher than the meteoblue ones (which from my experience are closer to reality)
![newplot (14) (1)](https://github.com/user-attachments/assets/e6aec993-8468-4568-b618-c9ada61ade4b)
I've checked the code for errors and tried with different models but I always get similar results 
![newplot (14) (2)](https://github.com/user-attachments/assets/55c3017d-8882-423e-9b07-a0d8cb57e5af)

Given your insider knowledge @patrick-zippenfenig do you have any idea what could be the difference? Do they just use a different irradiance threshold to compute sunshine hours or use a different definition altogether?

**Comment by patrick-zippenfenig** [comment]
Sunshine duration is calculated here:

https://github.com/open-meteo/open-meteo/blob/3e4c22c0b61919752c7a53d0e60ecb2a86b94f41/Sources/App/Helper/Solar/SunRiseSet.swift#L111-L183

The duration is defined as >120 W/m2 direct solar radiation on a normalised pane (DNI) by the WMO. This threshold is rather low and can be achieved even in cloudy conditions. I implemented it as close as possible to the official standard. 

Do you know any station measurement data for sunshine duration or DNI? The subjective perception might not be how "sunshine duration" is officially defined.

**Reply by guidocioni** [reply]
Yeah I know, I was just wondering if you knew what kind of definition meteoblue uses. 
Almost all stations with open data in Germany also measure sunshine duration, I could make a comparison

**Reply by patrick-zippenfenig** [reply]
Having some reference data would be great! Maybe the parameterisation can be shifted from `60-180` watts to `90-210` or `120-250` watts.

I would like to integrate satellite radiation data on a global scale that could be used to optimise solar radiation forecasts. I am also seeing larger differences in direct solar radiation forecasts from different models.

**Reply by guidocioni** [reply]
It's definitely possible to compare ICON-D2 to obs in Germany, but it may not yield meaningful results.
Anyway for the german stations you indeed have this data available with pretty good precision.

Is the computation done on the fly? In that case you could at least provide the thresholds as parameter in the API call (as done for some radiation vars if I'm not mistaken).

**Reply by patrick-zippenfenig** [reply]
Yes, it is done on the fly. The API also returns 15 minutely DNI which could then be evaluated against measured sunshine duration (DNI >120 watts/m2). Exposing a parameter to set the threshold might be a bit over-engineered.

**Comment by bonnells** [comment]
I have been accumulating data for a few months now and finally getting around to some analysis of the data. One area of interest is the sunshine calculations. I'm comparing the daily values as given and discussed here, with those that can be calculated using the hourly cloud %'s. Suppose the hourly total cloud amount is 70% then one could assume that it will be sunny for 30% of the hour, i.e 18 minutes. Using sunrise and sunset times and linearly interpolating values at these times, then doing a simple integration across the full day, allows for an alternative amount of 'sunshine' to be estimated. Comparing the two values results in considerable contrast. In some cases it is extreme, whereby the daily sunshine total amounts to > 10hrs [36000secs], yet the calculation based on the cloud amounts suggests no sunshine at all. I'll generate a plot of such comparisons to highlight these differences and post later. One experiment that Patrick might wish to consider is to [1] similarly generate sunshine values from the cloud %'s as above and [2] vary the solar threshold to see how the two match up. I'll have a look at his code and compare with mine [I did it in 'R' and will take a little while to digest] and see if I can do such an experiment myself. I am certain that the solar threshold is too high, but part of the problem may well be inherent within the models. I reckon they have higher solar values than actually occur. I have my own weather station, like countless others, and one thing I will do is verify the forecasts against these values. When I get something going, I'll pass on. Cheers, Dave.

**Comment by bonnells** [comment]
![meteo_sunshine_bonnells](https://github.com/user-attachments/assets/6b047e1f-b049-49d7-b2df-509298a037ec)
![CMA_sunshine_bonnells](https://github.com/user-attachments/assets/344ef06f-2b0c-487c-a550-992427f398a9)
![ICON_sunshine_bonnells](https://github.com/user-attachments/assets/1a068867-445d-4cf4-bbfc-ea7bf7fdf26c)
![BoM_sunshine_bonnells](https://github.com/user-attachments/assets/e4ae7e23-70df-435e-892a-ac62bfcccc2d)
![GFS_sunshine_bonnells](https://github.com/user-attachments/assets/040a67a6-d70d-4ddf-81ac-48a712c08e31)
![GEM_sunshine_bonnells](https://github.com/user-attachments/assets/db3b2e59-2835-49dc-b60f-38072bea8a0c)

I have quickly knocked up some plots in relation to my previous post. Will discuss/explore aspects of them later if any interest as only recently considered, however I expected what I got. The values calculated from the cloud data exceed those Patrick has derived on occasions, because I've used sunrise to sunset, and could easily be adjusted down to offset this by 'delaying' sunrise and 'bringing forward' sunset a few minutes. The data sets are virtually identical so differences between the plots highlight certain characteristics from each of the models, be it in the solar forecasts or the cloud forecasts. I'll consider other locations and see how they compare. I'm sure they'll largely agree. I've focused on locations within Australia, and the data here is where I live, 100km's to the north of Sydney. Cheers, Dave.

**Comment by guidocioni** [comment]
Hey @patrick-zippenfenig  I've done some quick exploration today.
Basically I took the historical forecast and reanalysis data from open-meteo and compared to satellite estimation of sunshine duration (purple line) and the observation of DWD for Hamburg (gray line). The latter is the ground truth and we can definitely trust the data from DWD. 
The results are quite interesting

![output](https://github.com/user-attachments/assets/5a0afbe5-07f2-4634-a80d-016a34c54d3d)

All the models seem to overestimate sunshine durations. GFS is, not surprisingly, completely off the charts, probably because it doesn't have any of the low-level cloud cover of the last week. `icon-d2` seems to be the best one, not surprisingly. 
Still I think the differences are only due to model biases because I checked in the documentation of the satellite dataset and they still use the same WMO threshold of 120 Wm2, and the satellite data are really close to the station ones.

<img width="655" alt="Screenshot 2024-11-09 at 15 38 58" src="https://github.com/user-attachments/assets/e0a7fed8-f26a-4865-a59d-5ac4d9de93ad">

So I'd say the computation of open meteo should be correct :)

**Reply by patrick-zippenfenig** [reply]
Great analysis! Of course it is just a single location for a limited time ;-) 

I am a bit tempted to use a more classical approach of using a cloud cover (20-80%) and multiply it with daylight duration to get sunshine hours, but it might produce similar results.

A better approach would be BIAS correction with satellite data to get better DNI values in the first place, but first I would have to implement global satellite radiation data, which I wanted to do at some point anyway....

**Reply by guidocioni** [reply]
The bias correction with satellite data is IMHO the best and only option, but it would be expensive and not sure if Worth it or not for the moment. In the end you want to be consistent and trust the data from the model.
If we start doing ML based bias correction then we may as well do it for every variable and not just for sunshine hours, but this a completely new model and scope 😂

**Comment by bonnells** [comment]
I am accumulating some analysis I am doing on sunshine hours and solar. I hope to pass some of this on in the near future.

      Verifying data comes from some data from the BoM. They have a page for each state, a couple of examples being:

      NSW: http://www.bom.gov.au/products/IDN65176.shtml

      QLD: http://www.bom.gov.au/products/IDQ60604.shtml

      There are only 6 locations with sunshine data that I have OpenMeteo data for comparison and around 3 dozen locations with solar data [across the whole country]. I've got around 8-9 months of data already. There are issues. I will expand on them with my results.

      One suggestion. Obtain data for your local grid points; i.e. where you live. Observe the sky on occasions and look at the amount of sun expected each hour/day and reflect on them. I actually have a personal weather station with solar data. It has its issues, but it offers further data to compare with the forecasts.
      If you do obtain some local data, compare the hourly forecasts for the cloud amounts and the hourly sunshine. There are instances of 100% cloud cover and 100% sunshine for the same hour. We need to look at ways to avoid this from happening.

      I am pleased to see some verification taking place. My focus is on verifying against actual observations and not the from the models. The reason will become obvious when I look at temperatures and rainfall.

      I hope we (and others) can look at ways to optimise the use of the resource that Patrick has made available to us.

      Cheers,

      Dave


________________________________

**Comment by patrick-zippenfenig** [comment]
Small updates:
- There was a bug in the CMA solar radiation which lead to wrong DNI values and therefore too high sunshine duration. This is now fixed.
- The SARAH-3 satellite radiation dataset is now available via the EUMETSAT API. I would like to integrate this dataset and use it to validate sunshine duration on a larger scale. This should help to identify model errors and errors in the sunshine duration calculation
